examples/lazy-input/main.rs

```diff
@@ -74,28 +74,34 @@ trait Db: salsa::Database {
     fn input(&self, path: PathBuf) -> Result<File>;
 }
 
-#[salsa::db]
-struct Database {
-    storage: salsa::Storage<Self>,
+struct LazyInput {
     logs: Mutex<Vec<String>>,
     files: DashMap<PathBuf, File>,
     file_watcher: Mutex<Debouncer<RecommendedWatcher>>,
 }
 
-impl Database {
+impl LazyInput {
     fn new(tx: Sender<DebounceEventResult>) -> Self {
-        let storage = Default::default();
         Self {
-            storage,
             logs: Default::default(),
             files: DashMap::new(),
             file_watcher: Mutex::new(new_debouncer(Duration::from_secs(1), tx).unwrap()),
         }
     }
 }
 
+impl UserData for LazyInput {
```

Comment:
What I understand is that the major change in this PR is to go from `Database` (application code) wrapping a `Storage` to `Database` (framework code) wrapping `UserData`. I don't understand the motivation for the change and I'm sure there are good reason for it. That's why my feedback here is primarily from a user perspective. 

To me, the old `Database` wraps a storage was easier to understand and use in the application code:

* I only need to name one unit, the database. With the new approach I have to both find a name for the `Database` alias AND the struct encapsulating `UserData`. I especially find naming `UserData` hard because it's just a collection of all the things not inside salsa. 
* I do find the name `UserData` confusing, especially when implementing in the application. Aren't all ingredients instances user data (it's not framework data)? 

I think this is probably mainly a naming problem, but I must admit that I fail to come up with a better name. Maybe `ExternalData`? But is the trait just about data? Isn't `salsa_event` an indicator that the trait does much more? 

I don't think it's blocking but we may want to improve the names of external concepts in follow up PRs.

Reply:
Thanks for the feedback. You are correct that this is the major change. I agree that the old approach was "nice" from a user point-of-view (which is why I pursued it), but it had some pretty strong disadvantages too. I'll write a separate comment on the motivations for doing the inversion, it'd be good to document them.

Regarding naming, I agree, `UserData` is kind of leaking my "salsa implementor" perspective. Perhaps `CustomData` or `UntrackedData`? I'm going to run with the latter.

I also think we need to document the pattern(s) we expect users to follow. In this PR I think there are three possibilities, from simplest to most complex:

* Simplest: if you don't need any custom fields, just use `DatabaseImpl::new` and `dyn Database` in your functions.
* More complex: If you have one set of data that you need across your entire application, you can define it in one place (perhaps a core crate) and then do `trait MyDatabase: Database`, `struct MyUntrackedData`, and implement `MyDatabase` for `DatabaseImpl<MyUntrackedData>` and `UntrackedData` for `MyUntrackedData.`
* Even more complex: If you have many crates and you don't want them all to depend on one base crate that defines your user data, you can make two traits, `MyDatabase1: Database` and `MyUntrackedData1: UntrackedData` and then implement `MyDatabase1` for all `DatabaseImpl<U>` where `U: MyUntrackedData1`. You can then define in some final crate a `MyUntrackedData` that implements `MyUntrackedData1` (along with `MyUntrackedData2..N` from other crates).

Overall I found it nicer to use for most tests (less boilerplate) but I agree that it's less "obvious' somehow. I'm thinking about it, maybe there's a way to reshuffle it.

It's *definitely* nice from the implementor perspective,_ because before it was always this "dance" where we didn't know the *actual* type of the database because it was embedded into something else.

Reply:
The more I think about it, the more I like `UntrackedData`. The idea then is that you instantiate a salsa database and it stores the tracked data; if you need other data, you can create it and get access to it yourself.

One other thought I had is that I think we could define a `DatabaseWith<U>` trait that you can use, so that the middle pattern (one but of untracked data) because like so:

* Simple: as above
* Medium: define your untracked data `MyUntrackedData` type in some core crate. Right all your code as `&dyn DatabaseWith<MyUntrackedData>` (you may want to create an alias), which is what I forced you to do before.
* Complex: define multiple untracked data pieces and code against them, basically unchanged.

Reply:
Oh, the last thought is that we can make `DatabaseImpl` crate-local, so you only work with a `Box<dyn Database>` or ` Box<dyn DatabaseWith>`, essentially. You create the database with `salsa::database()` and `salsa::database_with(u)`.

Reply:
(Hmm, maybe that doesn't *quite* work with the handle cloning...well, something like that.)

Reply:
The name `UntrackedData` eliminates most of my concerns. I think the name captures well what data is stored inside the struct. 

I have to play with the traits a bit to get a better feeling on how complicated they feel. Our setup currently uses multiple databases but we may get away with a single database and just two untracked (one for production and one for testing) the more a I think about it.

Reply:
> This use case now feels significantly more complicated because it

It sounds as if I feel strongly about it. I don't. It's just something I noticed, and I strongly favor better performance over avoiding boilerplate at all costs (when setting up the database) . 

I haven't yet figured out how to implement `Upcast` with the new design. Could we use the `views` API instead?

Reply:
Resolved and switched it over. I think you should be able to do whatever you were doing before.